packages/frames.js/src/core/middleware/openframes.ts

```diff
@@ -0,0 +1,107 @@
+import { ClientProtocolId } from "../..";
+import { RequestBodyNotJSONError } from "../errors";
+import { FramesMiddleware, FramesMiddlewareReturnType } from "../types";
+import { isFrameDefinition } from "../utils";
+
+type FramesMessageContext = {
+  // TODO: this should extend a base FrameMessage with common fields
+  message?: any;
+  clientProtocol?: string;
+};
+
+async function cloneJsonBody(request: Request): Promise<JSON | undefined> {
+  try {
+    // use clone just in case someone wants to read body somewhere along the way
+    const body = await request
+      .clone()
+      .json()
+      .catch(() => {
+        throw new RequestBodyNotJSONError();
+      });
+
+    return body;
+  } catch (e) {
+    if (e instanceof RequestBodyNotJSONError) {
+      return undefined;
+    }
+
+    console.error(e);
+
+    return undefined;
+  }
+}
+
+async function nextInjectClientProtocols({
+  nextResult,
+  accepts,
+}: {
+  nextResult: FramesMiddlewareReturnType;
+  accepts: ClientProtocolId[];
+}) {
+  const result = await nextResult;
+  if (isFrameDefinition(result)) {
+    return { ...result, accepts };
+  }
+  return result;
+}
+
+export function openframes({
+  clientProtocolHandlers: handlers,
+}: {
+  /**
+   * Validators and handlers for frame messages from different client protocols.
+   * Key is the client protocol ID (e.g. "id@version")
+   * Value is an object containing a frame message validator and handler
+   * The output of the handler is added to the context as `message`
+   */
+  clientProtocolHandlers: Record<
+    string,
+    {
+      // TODO: better types
```

Comment:
automatic types from middleware is going to be really helpful. Perhaps @michalkvasnicak can take a look once this is merged?

Reply:
We could afterwards.